src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_comm.py

```diff
@@ -199,18 +199,12 @@ def __init__(self, sock, terminate_on_socket_close=True):
         self.setName("pydevd.Reader")
         self.process_net_command = process_net_command
         self.process_net_command_json = PyDevJsonCommandProcessor(self._from_json).process_net_command_json
-        self.global_debugger_holder = GlobalDebuggerHolder
-
-        self._dap_messages_listeners = []
 
     def _from_json(self, json_msg, update_ids_from_dap=False):
         return pydevd_base_schema.from_json(json_msg, update_ids_from_dap, on_dict_loaded=self._on_dict_loaded)
 
-    def add_dap_messages_listener(self, dap_commands_listener):
-        self._dap_messages_listeners.append(dap_commands_listener)
-
     def _on_dict_loaded(self, dct):
-        for listener in self._dap_messages_listeners:
+        for listener in GlobalDebuggerHolder.global_dbg.dap_messages_listeners:
```

Comment:
It should use `self.global_debugger_holder` as was previously in the code... the reason for this is that module globals become `None` during interpreter shutdown, so, we need to keep a reference.

Reading the code it seems that the `PyDB` could actually be passed in the constructor (that would make the dependency more explicit too), so, I think the best here would be to pass `py_db` in both constructors (`ReaderThread` and `WriterThread`), store the reference as `self._py_db` and use that instance instead of `GlobalDebuggerHolder.global_dbg`.